### ISSUE TITLE
question: about encodeURI. server.js:181 (@3c3b47e)

### ISSUE THREAD

**Issue by Woodya** [prerender/prerender#170]
Todd, I'm curious why this "extra" encoding is here.

I've been tacking down an issue where inbound comma, colon etc. are turned into %2C (or appropriate) by util.getUr(), and then this is turned into %252C by the additional
```var uri = encodeURI(req.prerender.url.replace(/%20/g, ' '));```

i'm curious what this is all about? i've had a look at #121 .. and i removed the extra encodeURI and the utf8 characters are seen correctly by the js on my page (not when encoded twice of course).

Could you explain what drove the addition of this encode step, andy why it may be unsafe/erroneous for me to remove it.
thanks

**Comment by thoop** [comment]
Can you give an example of a URL that causes this? Or can you email me the URL you're testing and having problems with? I'd like to just test it and see where util.getUrl() is encoding it because I'm not able to reproduce that locally. todd@prerender.io

And what version of node are you running?

**Comment by Woodya** [comment]
example.com/?foo=1,2,3

**Comment by thoop** [comment]
I'm on v0.10.33

I put this line:
`console.log('encoded URI:', uri);`
before `req.prerender.page.open(uri...`:

I tried both of these and get the same output on my console:
`curl http://localhost:1338/http://example.com/?foo=1,2,3`
`curl http://localhost:1338/http://example.com/?foo=1%2C2%2C3`

```
2015-03-14T00:40:24.333Z getting http://example.com/?foo=1,2,3
encoded URI: http://example.com/?foo=1,2,3
2015-03-14T00:40:24.956Z got 200 in 623ms for http://example.com/?foo=1,2,3
```
Are you seeing something different?

**Comment by Woodya** [comment]
yes. i suspect you did not start with &_escaped_fragment_= in your url when triggering pre-render (this is the hint to my nginx config to proxy through my local prerender).
i should have made that explicit.
it when prerender encounters the _escaped_fragment_ url/hash parameter it trims it from the query dict, and then deletes the search attr of the parsed url object (in util.getUrl()), then url.format(parts) is called, which will encode the various parts of the search parameters when reconstructing the url. Thus, the url has already been encoded in this path.

i put these in the analogous location:
```
            console.log('OPEN', req.prerender.url);
            console.log('open', encodeURI(req.prerender.url.replace(/%20/g, ' '))); 
```

resulting in:
```
OPEN https://localhost.zumper.com/?foo=1%2C2%2C3
open https://localhost.zumper.com/?foo=1%252C2%252C3
```

**Comment by thoop** [comment]
Ah I see, thanks for that. Let me run through that again.

**Comment by Woodya** [comment]
@thoop I've been running tests without the ```encodeURI``` and everything seems good. including urls with encodeable characters.

what was the reason for this line of code to begin with? (the /%20/g snippet smells of 'quick fixing some bug, possibly with phantom')

**Comment by thoop** [comment]
Sorry I haven't gotten to this yet. I still want to check this. The %20 fix was something where phantomjs 1.9.x with a %20 in the URL was (sometimes) causing angularjs to go into an infinite loop.

I'll check it soon and get back to you. I just want to make sure on our end that we don't break anything before changing that on master since we process millions of URLs.

**Comment by damiengenet** [comment]
Same problem here, the encodeURI breaks AngularJS routing (with infinite digest loop errors) when there are non ascii characters in the url (tested for accents and Korean characters). Why was the encode/decode initially necessary here?

**Comment by thoop** [comment]
If you remove the encode, does it work with accents and korean characters as well? Sorry I haven't gotten to this yet. Been insanely busy...

**Comment by damiengenet** [comment]
No problem :).

After further testing, removing the encodeURI fixes most extended characters (we randomly tested 2 bytes and 3 bytes utf-8 characters in several blocks), except for the most used part of the latin1 block (not all characters there strangely) and space which still doesn't work. So it's actually not a usable fix.

Changing to PhantomJS 2.0 (without encode) seems to fix all problems for us. Quickly looking at PhantomJS 2 source code it seems like they changed their handling of the url character encoding internally. However this still requires quite a bit of tweaking as they still don't have any proper build for Linux and as such most of the ecosystem is not ready yet and still based on older 1.9 versions.

**Comment by thoop** [comment]
We are using PhantomJS 2.0 in production. Your notes on PhantomJS 1.9.8 problems when removing encodeURI is why we were using the encodeURI with PhantomJS 1.9.8.

I just pushed a fix for this in version `3.15.2`. Since Prerender defaults to PhantomJS 1.9.8, you can set `export DISABLE_URI_ENCODE=true` to turn off the encoding. It defaults to being left on for now.

**Comment by damiengenet** [comment]
encodeURI didn't work for latin1 characters for us either, not sure why. We also had to move to PhantomJS 2.0 in the end. Thanks for your help

**Comment by thoop** [comment]
Should be fixed in our latest version of Prerender (v4.0.3).
